session.go

```diff
@@ -0,0 +1,233 @@
+package apns
+
+import (
+	"container/list"
+	"errors"
+	"io"
+	"sync"
+)
+
+type SessionError struct {
+	Notification Notification
+	Err          Error
+}
+
+func (s SessionError) Error() string {
+	return s.Err.Error()
+}
+
+type Session interface {
+	Send(n Notification) error
+	Connect() error
+	RequeueableNotifications() []Notification
+	Disconnect()
+	Disconnected() bool
+}
+
+type buffer struct {
+	size int
+	m    sync.Mutex
+	*list.List
+}
+
+func newBuffer(size int) *buffer {
+	return &buffer{size, sync.Mutex{}, list.New()}
+}
+
+func (b *buffer) Add(v interface{}) *list.Element {
+	b.m.Lock()
+	defer b.m.Unlock()
+
+	e := b.PushBack(v)
+
+	if b.Len() > b.size {
+		b.Remove(b.Front())
+	}
+
+	return e
+}
+
+type sessionState int
+
+const (
+	sessionStateNew          sessionState = iota
+	sessionStateConnected    sessionState = iota
+	sessionStateDisconnected sessionState = iota
+)
+
+type session struct {
+	b *buffer
+
+	conn  Conn
+	connm sync.Mutex
+
+	st  sessionState
+	stm sync.Mutex
+
+	id  uint32
+	idm sync.Mutex
+
+	err SessionError
+}
+
+func NewSession(conn Conn) Session {
+	return &session{
+		st:    sessionStateNew,
+		stm:   sync.Mutex{},
+		conn:  conn,
+		connm: sync.Mutex{},
+		idm:   sync.Mutex{},
+		b:     newBuffer(50),
+	}
+}
+
+func (s *session) Connect() error {
+	if s.isNew() {
+		return errors.New("can't connect unless the session is new")
+	}
+
+	go s.readErrors()
+	return nil
+}
+
+func (s *session) isNew() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st != sessionStateNew
+}
+
+func (s *session) Disconnected() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st == sessionStateDisconnected
+}
+
+func (s *session) Connnected() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st == sessionStateConnected
+}
+
+func (s *session) Send(n Notification) error {
+	// If disconnected, error out
+	if s.Connnected() {
```

Comment:
This code doesn't read like the comment above.

Should it be `!s.Connected()` or `s.Disconnected()`